### DISCUSSION TITLE
Help with generating my dsl

### DISCUSSION THREAD

**Discussion by Anthie123** [eclipse-langium/langium#1198] [Q&A] [answered]
Hi,
I have implemented a simple dsl where I had to fix the references with help of the community (see: [my former issue](https://github.com/eclipse-langium/langium/discussions/1194) ). 

Now I have implemented the [builtIn Library](https://langium.org/guides/builtin-library/) feature described in the documentation and this works correctly in the VS Code extension. But when I start the generation of my language the references are not resolved and I get validation errors. 

I have seen the solution of https://github.com/eclipse-langium/langium/discussions/1188 but I have no overridden implementation of the DocumentBuilder and using this workaround leads me to problems with esm and commonjs in vscode modules.

Is there some easy workaround besides using a nightly build of langium?

Best regards
Andrej

**Comment by msujew** [comment]
I think the `DocumentBuilder` override is the easiest solution to the issue for adopters. I have a bit more time next week, and we will probably release a 2.1.0 (maybe a 1.4.0 as well) that includes https://github.com/eclipse-langium/langium/pull/1190 (among others).

**Reply by Anthie123** [reply]
I tried to override the DocumentBuilder. I registered my version in the SharedModule  under the workspace, overode the build and update method and introduced the emitUpdate method similar to your commit.

But I get problems with the esm and commonjs now: 
`
$ ./bin/cli.js generate Simple.testdsl 
.../testdsl/node_modules/vscode-languageclient/lib/node/main.js:25
const vscode_1 = require("vscode");
                 ^

Error [ERR_REQUIRE_ESM]: require() of ES Module .../testdsl/node_modules/vscode/dist/api.js from .../testdsl/node_modules/vscode-languageclient/lib/node/main.js not supported.
Instead change the require of api.js in .../testdsl/node_modules/vscode-languageclient/lib/node/main.js to a dynamic import() which is available in all CommonJS modules.
    at Object.<anonymous> (.../testdsl/node_modules/vscode-languageclient/lib/node/main.js:25:18)
    at async Promise.all (index 0) {
  code: 'ERR_REQUIRE_ESM'
}
`
This happens in the extension and the generation. How could I solve this problem which is not in my code?

**Reply by msujew** [reply]
You probably accidentally imported `vscode-languageclient` in your language server code. The `vscode-languageclient` library tries to import `vscode`, which is not a valid import outside of the vscode extension host process. This should be fixable by replacing the imports of `vscode-languageclient` with `vscode-languageserver`.

**Reply by Anthie123** [reply]
Hi Mark,
thank you for your answer. You were right, the "CancellationToken" was imported from the wrong module. The code is running but the references are still not resolved with the DocumentBuilder-Fix. 
```
$ ./bin/cli.js generate Simple.testdsl 
There are validation errors:
line 2: Could not resolve reference to TestCase named 'LoginPage'. [LoginPage]
line 3: Could not resolve reference to FunctionDefinition named 'fillLogin'. [fillLogin]
line 4: Could not resolve reference to FunctionDefinition named 'login'. [login]
line 8: Could not resolve reference to TestCase named 'NamePage'. [NamePage]
line 9: Could not resolve reference to FunctionDefinition named 'fillName'. [fillName]
line 10: Could not resolve reference to FunctionDefinition named 'process'. [process]
line 11: Could not resolve reference to FunctionDefinition named 'checkName'. [checkName]
```
Do you have any idea, where the problem could be?

**Reply by msujew** [reply]
Without looking into your code, I can't really tell anything. Can you share some of the code that you use to load the text files and how you overrode the DocumentBuilder?

**Reply by Anthie123** [reply]
The overridden DocumentBuilder:
```
import { AstNode, BuildOptions, DefaultDocumentBuilder, DocumentState, LangiumDocument, LangiumSharedServices, URI, ValidationCategory, interruptAndCheck, stream } from "langium";
import { CancellationToken } from "vscode-languageserver";


export type DocumentUpdateListener = (changed: URI[], deleted: URI[]) => void | Promise<void>

export class TestdslDocumentBuilder extends DefaultDocumentBuilder {
    constructor(services: LangiumSharedServices) {
        super(services);
    }

    override async build<T extends AstNode>(documents: Array<LangiumDocument<T>>, options: BuildOptions = {}, cancelToken = CancellationToken.None): Promise<void> {
...
```
The methods are just copied from your code.

The Module:
```
...
export type TestdslSharedServices = LangiumSharedServices;

export const TestdslSharedModule: Module<TestdslSharedServices, DeepPartial<TestdslSharedServices>> = {
    workspace: {
        WorkspaceManager: (services) => new TestdslWorkspaceManager(services),
        DocumentBuilder: (services) => new TestdslDocumentBuilder(services),
    }
}
...
```

The Workspace for the "Builtin Lib":
```
import {
    AstNode,
    DefaultWorkspaceManager,
    LangiumDocument,
    LangiumDocumentFactory,
    LangiumSharedServices
} from "langium";
import { WorkspaceFolder } from 'vscode-languageserver';
import vscodeUri  from "vscode-uri";
import { testdslLib } from "../buildinLib/testdslLib.js";


export class TestdslWorkspaceManager extends DefaultWorkspaceManager {

    private documentFactory: LangiumDocumentFactory;

    constructor(services: LangiumSharedServices) {
        super(services);
        this.documentFactory = services.workspace.LangiumDocumentFactory;
    }

    protected override async loadAdditionalDocuments(
        folders: WorkspaceFolder[],
        collector: (document: LangiumDocument<AstNode>) => void
    ): Promise<void> {
        await super.loadAdditionalDocuments(folders, collector);
        // Load our library using the `builtin` URI schema
        collector(this.documentFactory.fromString(testdslLib, vscodeUri.URI.parse('builtin:///testdslLib.testdsl')));
    }
}
```
The testdsl-Import is just a string with my dsl code.

The other code is just like in the hello-world example generated by yeoman. Only exception was, that I had to modify the call in your cli.js in the bin folder:
```
import cli from '../out/cli/main.js';

cli();
```
and import the require function in the src/cli/main.ts to make the program.version(...) method call work.
```
...
import { createRequire } from "module";

const require = createRequire(import.meta.url);
...
program
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        .version(require('../../package.json').version);
...
```
I hope this helps for diagnostic.

**Reply by cdietrich** [reply, marked as answer]
to me it looks like default cli-util does not use`WorkspaceManager.initializeWorkspace`

**Reply by cdietrich** [reply]
which reminds me what i did some half a year ago here: https://github.com/cdietrich/langium-1x-experiments/blob/b360d439cf736beee001de07ac18b2b4c88c6858/src/cli/generator.ts#L11

**Reply by msujew** [reply]
> to me it looks like default cli-util does not useWorkspaceManager.initializeWorkspace

Right, @Anthie123 are you actually calling `initializeWorkspace` in your cli?

**Reply by Anthie123** [reply]
I have tried this and now it seems to work. Thank you very much for your support. Especially cool to hear again from you Christian. Would be great to meet you again.
